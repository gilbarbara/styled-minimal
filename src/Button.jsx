import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { withComponent } from './utils/helpers';
import { buttonType, sizeTypeFull, variantType } from './utils/propTypes';
import { ButtonStyles } from './utils/system';

import { StyledBadge } from './Badge';

export const StyledButton = styled.button`
  ${ButtonStyles.base};
  
  &:disabled {
    pointer-events: none;
  }
  
  &:hover {
    transition: transform 0.2s;
    transform: scale(1.05);
  }
  
  &:focus {  
      ${ButtonStyles.outlineColor};
  }
  
  ${StyledBadge} {
    margin-left: 5px;
  }
`;

StyledButton.propTypes = {
  /** boolean indicating whether the button should render with an animation */
  animate: PropTypes.bool,
  as: PropTypes.string,
  /** match its parent width */
  block: PropTypes.bool,
  children: PropTypes.node.isRequired,
  /** boolean indicating whether the button should render as disabled */
  disabled: PropTypes.bool,
  /** callback on the click event */
  onClick: PropTypes.func,
  outline: PropTypes.bool,
  size: sizeTypeFull,
  type: buttonType,
  variant: variantType,
};

StyledButton.defaultProps = {
  animate: false,
  block: false,
  disabled: false,
  outline: false,
  size: 'md',
  type: 'button',
  variant: 'primary',
};

export default withComponent(StyledButton, 'Button');
