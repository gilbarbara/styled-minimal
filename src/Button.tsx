import * as React from 'react';
import styled, { css } from 'styled-components';

import { getColor, getTheme, px } from './utils/helpers';
import { baseStyles, sx, system, variantStyles } from './utils/system';

import { StyledBadge } from './Badge';

import { ButtonProps } from './types';

export const StyledButton = styled('button')<ButtonProps>(
  props => {
    const { block, busy, invert, size = 'md' } = props;
    const { borderRadius, disabledOpacity, fontSize, lineHeight, loader, padding } = getTheme(
      props,
      'button',
    );

    return css`
      ${baseStyles};
      ${variantStyles};
      ${busy ? loader(invert ? '#ccc' : '#fff') : ''};
      align-items: center;
      border-radius: ${px(borderRadius[size])};
      box-shadow: none;
      cursor: pointer;
      display: inline-flex;
      font-family: inherit;
      font-size: ${px(fontSize[size])};
      line-height: ${lineHeight};
      padding: ${px(padding[size][0])} ${px(padding[size][1])};
      text-decoration: none;
      width: ${block ? '100%' : 'auto'};

      &:disabled {
        opacity: ${disabledOpacity};
        pointer-events: none;
      }

      &:focus {
        outline-color: ${getColor};
      }

      ${StyledBadge} {
        margin-left: 5px;
      }
    `;
  },
  sx,
  system,
);

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => (
  <StyledButton ref={ref} {...props} />
));

Button.defaultProps = {
  block: false,
  busy: false,
  dark: false,
  disabled: false,
  invert: false,
  size: 'md',
  type: 'button',
  variant: 'primary',
};
