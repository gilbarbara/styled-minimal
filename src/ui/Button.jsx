import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { getProp } from '../utils/helpers';
import { colors } from '../utils/styles';

const Wrapper = styled.button.attrs({
  type: 'button',
})`
  align-items: center;
  border-radius: ${getProp('buttonRadius', 'size', 'md')};
  background-color: ${getProp('palette', 'variant', 'primary')};
  border: 0;
  color: ${({ variant }) => (variant === 'default' ? colors.black : colors.white)};
  display: flex;
  font-size: ${getProp('fontSizes', 'size', 'md')};
  font-weight: 700;
  padding: ${getProp('buttonPadding', 'size', 'md')};
  text-transform: uppercase;
  ${({ loading }) => (loading ? getProp('buttonLoader') : '')}
  
  [class^=i-] {
    margin-right: 1rem;
  }
`;

const Button = ({ children, ...props }) => (
  <Wrapper {...props}>{children}</Wrapper>
);

Button.propTypes = {
  children: PropTypes.node.isRequired,
  loading: PropTypes.bool,
  size: PropTypes.string,
  variant: PropTypes.string,
};

Button.defaultProps = {
  size: 'md',
  variant: 'primary',
};

export default Button;
