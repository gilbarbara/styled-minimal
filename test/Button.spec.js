import React from 'react';
import { mount } from 'enzyme';

import Button from '../src/Button';

describe('Button', () => {
  const wrapper = mount(
    <Button>CLICK ME</Button>
  );

  it('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
