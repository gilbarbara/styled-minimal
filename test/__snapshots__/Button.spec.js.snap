// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render properly 1`] = `
.c0 {
  box-sizing: border-box;
  background-color: #2E2E2E;
  border: 1px solid #2E2E2E;
  color: #fff;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0px;
  box-shadow: none;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: inherit;
  font-size: 16px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1;
  padding: 9px 16px;
  -webkit-text-decoration: none;
  text-decoration: none;
  width: auto;
}

.c0:disabled {
  pointer-events: none;
  opacity: 0.7;
}

.c0:focus {
  outline-color: #2E2E2E;
}

.c0 .c1 {
  margin-left: 5px;
}

<Button
  animate={false}
  as="button"
  block={false}
  disabled={false}
  outline={false}
  size="md"
  type="button"
  variant="primary"
>
  <button
    className="c0"
    disabled={false}
    size="md"
    type="button"
  >
    CLICK ME
  </button>
</Button>
`;
